Double write during publish

Publishing a self-contained WPF template app writes some files twice because `ResolvedFileToPublish` has the same dll from different runtime packs (for example, Microsoft.Win32.Registry.dll comes from both netcoreapp and the windowsdesktop pack).
This causes crossgen failures when used together with `PublishReadyToRun=true`.

Looks very similar to https://github.com/dotnet/sdk/issues/3035 which was addressed in https://github.com/dotnet/sdk/pull/3021, but I'm still seeing this with version `3.0.100-preview6-012031`.

/cc @fadimounir @nguerrera @peterhuene 

edit: it fails during `PublishReadyToRun=true` only when used together with the linker. Without linking, the duplicates are already crossgen'd, and so they don't hit the failure. In any case, `ResolvedFileToPublish` has duplicates that shouldn't be there, and we end up with `DoubleWrites` of some files.

<img width="565" alt="image" src="https://user-images.githubusercontent.com/787361/58192612-00a04780-7c76-11e9-8ffc-102d4b993fb2.png">

We now have folks hitting this when trying to trim+R2R their WPF apps, FYI:
```
"D:\dev\NuGetPackageExplorer\PackageExplorer\NuGetPackageExplorer.csproj" (publish target) (1) ->
(_CreateR2RImages target) ->
  C:\Program Files\dotnet\sdk\3.0.100-preview6-012151\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(284,28): error MSB4094: "obj\Debug\ netcoreapp3.0\win-x86\linked\Microsoft.Win32.Registry.dll;obj\Debug\netcoreapp3.0\win-x86\linked\Microsoft.Win32.Registry.dll" is an invalid value for  the "CompilationEntry" parameter of the "RunReadyToRunCompiler" task. Multiple items cannot be passed into a parameter of type "Microsoft.Build.Frame work.ITaskItem". [D:\dev\NuGetPackageExplorer\PackageExplorer\NuGetPackageExplorer.csproj]
```
@nguerrera

--- a/src/Tasks/Common/ConflictResolution/ConflictResolver.cs
+++ b/src/Tasks/Common/ConflictResolution/ConflictResolver.cs
@@ -349,6 +349,25 @@ namespace Microsoft.NET.Build.Tasks.ConflictResolution
                 return item2;
             }
 
+            if (item1.ItemType == ConflictItemType.CopyLocal && item2.ItemType == ConflictItemType.CopyLocal)
+            {
+                // If two items are copy local, we must pick one even if versions are identical, as only 
+                // one of them can be copied locally. The policy here must be deterministic, but it can
+                // be chosen arbitrarily. The assumption is that the assemblies are fully semantically 
+                // equivalent.
+                //
+                // We choose ordinal string comparison of package id as a final tie-breaker for this case. 
+                // We will get here in the real case of frameworks with overlapping assemblies (including 
+                // version) and self-contained apps. The assembly we choose here is not guaranteed to match
+                // the assembly that would be chosen by the host for a framework-dependent app. The host
+                // is free to make its own deterministic but arbitrary choice.
+                int cmp = string.CompareOrdinal(item1.PackageId, item2.PackageId);
+                if (cmp != 0)
+                {
+                    return cmp < 0 ? item1 : item2;
+                }
+            }
+
             if (logUnresolvedConflicts)
             {
                 string message = conflictMessage + SENTENCE_SPACING + string.Format(CultureInfo.CurrentCulture, Strings.ConflictCouldNotDetermineWinner);

--- a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets
@@ -553,8 +553,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <ItemGroup>
       <_ResolvedCopyLocalPublishAssets Include="@(ReferenceCopyLocalPaths)"
                                        Exclude="@(_ResolvedCopyLocalBuildAssets);@(RuntimePackAsset)"
-                                       Condition="'$(PublishReferencesDocumentationFiles)' == 'true' or '%(Extension)' != '.xml'">
-        <DestinationSubPath>%(ReferenceCopyLocalPaths.DestinationSubDirectory)%(Filename)%(Extension)</DestinationSubPath>
+                                       Condition="'$(PublishReferencesDocumentationFiles)' == 'true' or '%(ReferenceCopyLocalPaths.Extension)' != '.xml'">
+        <DestinationSubPath>%(ReferenceCopyLocalPaths.DestinationSubDirectory)%(ReferenceCopyLocalPaths.Filename)%(ReferenceCopyLocalPaths.Extension)</DestinationSubPath>
       </_ResolvedCopyLocalPublishAssets>
     </ItemGroup>
 

--- a/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAHelloWorldProject.cs
+++ b/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAHelloWorldProject.cs
@@ -501,8 +501,11 @@ public static class Program
                 .HaveStdOutContaining("NETSDK1085");
         }
 
+        [WindowsOnlyFact]
+        public void It_contains_no_duplicates_in_resolved_publish_assets_on_windows()
+            => It_contains_no_duplicates_in_resolved_publish_assets("windows");
+
         [Theory]
-        [InlineData("windows")]
         [InlineData("console")]
         [InlineData("web")]
         public void It_contains_no_duplicates_in_resolved_publish_assets(string type)
@@ -520,10 +523,6 @@ public static class Program
             switch (type)
             {
                 case "windows":
-                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-                    {
-                        return; // Cannot build for windows on non-Windows
-                    }
                     testProject.ProjectSdk = "Microsoft.NET.Sdk.WindowsDesktop";
                     testProject.AdditionalProperties.Add("UseWpf", "true");
                     testProject.AdditionalProperties.Add("UseWindowsForms", "true");

--- a/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAHelloWorldProject.cs
+++ b/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAHelloWorldProject.cs
@@ -501,8 +501,11 @@ public static class Program
                 .HaveStdOutContaining("NETSDK1085");
         }
 
-        [Fact]
-        public void It_contains_no_duplicates_in_resolved_publish_assets()
+        [Theory]
+        [InlineData("windows")]
+        [InlineData("console")]
+        [InlineData("web")]
+        public void It_contains_no_duplicates_in_resolved_publish_assets(string type)
         {
             // Use a specific RID to guarantee a consistent set of assets
             var testProject = new TestProject()
@@ -514,9 +517,29 @@ public static class Program
                 IsExe = true
             };
 
+            switch (type)
+            {
+                case "windows":
+                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    {
+                        return; // Cannot build for windows on non-Windows
+                    }
+                    testProject.ProjectSdk = "Microsoft.NET.Sdk.WindowsDesktop";
+                    testProject.AdditionalProperties.Add("UseWpf", "true");
+                    testProject.AdditionalProperties.Add("UseWindowsForms", "true");
+                    break;
+               case "console":
+                    break;
+                case "web":
+                    testProject.ProjectSdk = "Microsoft.NET.Sdk.Web";
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(type));
+            }
+
             testProject.PackageReferences.Add(new TestPackageReference("NewtonSoft.Json", "9.0.1"));
 
-            var testAsset = _testAssetsManager.CreateTestProject(testProject, testProject.Name)
+            var testAsset = _testAssetsManager.CreateTestProject(testProject, testProject.Name, identifier: type)
                 .WithProjectChanges(project =>
                 {
                     project.Root.Add(XElement.Parse(@"
@@ -526,12 +549,12 @@ public static class Program
     </RemoveDuplicates>
     <Message Condition=""'@(_ResolvedCopyLocalPublishAssets)' != '@(FilteredAssets)'"" Importance=""High"" Text=""Duplicate items are present in: @(_ResolvedCopyLocalPublishAssets)!"" />
     <ItemGroup>
-        <AssetFilenames Include=""@(_ResolvedCopyLocalPublishAssets->'%(Filename)%(Extension)')"" />
+        <AssetDestinationSubPaths Include=""@(_ResolvedCopyLocalPublishAssets->'%(DestinationSubPath)')"" />
     </ItemGroup>
-    <RemoveDuplicates Inputs=""@(AssetFilenames)"">
-        <Output TaskParameter=""Filtered"" ItemName=""FilteredAssetFilenames""/>
+    <RemoveDuplicates Inputs=""@(AssetDestinationSubPaths)"">
+        <Output TaskParameter=""Filtered"" ItemName=""FilteredAssetDestinationSubPaths""/>
     </RemoveDuplicates>
-    <Message Condition=""'@(AssetFilenames)' != '@(FilteredAssetFilenames)'"" Importance=""High"" Text=""Duplicate filenames are present in: @(_ResolvedCopyLocalPublishAssets)!"" />
+    <Message Condition=""'@(AssetDestinationSubPaths)' != '@(FilteredAssetDestinationSubPaths)'"" Importance=""High"" Text=""Duplicate DestinationSubPaths are present in: @(AssetDestinationSubPaths)!"" />
 </Target>"));
                 })
                 .Restore(Log, testProject.Name);

